app/src/ui/dialog/dialog.tsx

```diff
@@ -221,9 +220,135 @@ export class Dialog extends React.Component<IDialogProps, IDialogState> {
     this.setState({ isAppearing: true })
     this.scheduleDismissGraceTimeout()
 
+    this.focusFirstSuitableChild()
+
     window.addEventListener('focus', this.onWindowFocus)
   }
 
+  /**
+   * Attempts to move keyboard focus to the first _suitable_ child of the
+   * dialog.
+   *
+   * The original motivation for this function is that while the order of the
+   * Ok, and Cancel buttons differ between platforms (see OkCancelButtonGroup)
+   * we don't want to accidentally put keyboard focus on the destructive
+   * button (like the Ok button in the discard changes dialog) but rather
+   * on the non-destructive action. This logic originates from the macOS
+   * human interface guidelines
+   *
+   * From https://developer.apple.com/design/human-interface-guidelines/macos/windows-and-views/dialogs/:
+   *
+   *   "Users sometimes press Return merely to dismiss a dialog, without
+   *   reading its content, so it’s crucial that a default button initiate
+   *   a harmless action. [...] when a dialog may result in a destructive
+   *   action, Cancel can be set as the default button."
+   *
+   * The same guidelines also has this to say about focus:
+   *
+   *   "Set the initial focus to the first location that accepts user input.
+   *    Doing so lets the user begin entering data immediately, without needing
+   *    to click a specific item like a text field or list."
+   *
+   * In attempting to follow the guidelines outlined above we follow a priority
+   * order in determining the first suitable child.
+   *
+   *  1. The element with the lowest positive tabIndex
+   *     This might sound counterintuitive but imagine the following pseudo
+   *     dialog this would be button D as button D would be the first button
+   *     to get focused when hitting Tab.
+   *
+   *     <dialog>
+   *      <button>A</button>
+   *      <button tabIndex=3>B</button>
+   *      <button tabIndex=2>C</button>
+   *      <button tabIndex=1>D</button>
+   *     </dialog>
+   *
+   *  2. The first element which is either implicitly keyboard focusable (like a
+   *     text input field) or explicitly focusable through tabIndex=0 (like a TabBar
+   *     tab)
+   *
+   *  3. The first submit button. We use this as a proxy for what macOS HIG calls
+   *     "default button". It's not the same thing but for our purposes it's close
+   *     enough.
+   *
+   *  4. Any remaining button
+   *
+   */
+  private focusFirstSuitableChild() {
```

Comment:
i won't pretend to fully understand the internal logic of this method, but i'm a bit concerned that we're always overriding default focus behavior with it. is there a way we could only use this logic if the dialog is destructive?

Reply:
The intention is to always override the default focus behavior because the default logic doesn't fit our needs. The diff has a number of examples of where we've now been able to remove explicit focus calls by individual dialogs to work around shortcomings in the default focus logic (`acknowledgements.tsx` is a good example). With this we've got a consistent focus logic (albeit custom) that should hopefully suit our needs. Should we discover that a particular dialog suffers from this focus logic (which is certainly possible, I've tested a lot of them but it's possible I missed something) I think we should either see if we can fix the logic in such a way that it works for all dialogs and if that's not possible we should add an opt-out prop for the dialog component that can be explicitly set for that case.

All that said dialogs still have the ability to control focus manually after they've been mounted.